src/ILVerify/src/ILImporter.StackValue.cs

```diff
@@ -165,194 +174,192 @@ public override string ToString()
+                return null;
+
+            switch (verificationType.Category)
+            {
                 case TypeFlags.SByte:
                 case TypeFlags.Int16:
                 case TypeFlags.Int32:
-                case TypeFlags.Int64:
-                case TypeFlags.IntPtr:
-                case TypeFlags.Boolean:
-                case TypeFlags.Char:
+                    return type.Context.GetWellKnownType(WellKnownType.Int32);
                 case TypeFlags.Single:
                 case TypeFlags.Double:
-                    return category;
+                    return type.Context.GetWellKnownType(WellKnownType.Double);
+                default:
+                    return verificationType;
             }
-
-            return TypeFlags.Unknown;
         }
 
         static bool IsSameReducedType(TypeDesc src, TypeDesc dst)
         {
-            var srcCategory = GetReducedTypeCategory(src);
-            if (srcCategory == TypeFlags.Unknown)
-                return false;
-            return srcCategory == GetReducedTypeCategory(dst);
+            return GetReducedType(src) == GetReducedType(dst);
         }
 
-        bool IsAssignable(TypeDesc src, TypeDesc dst, bool allowSizeEquivalence = false)
+        static bool IsSameVerificationType(TypeDesc src, TypeDesc dst)
         {
-            if (src == dst)
-                return true;
-
-            if (src.IsValueType || dst.IsValueType)
-            {
-                if (allowSizeEquivalence && IsSameReducedType(src, dst))
-                    return true;
-
-                // TODO IsEquivalent
-                return false;
-            }
-
-            return CastingHelper.CanCastTo(src, dst);
+            return GetVerificationType(src) == GetVerificationType(dst);
         }
 
-        bool IsAssignable(StackValue src, StackValue dst)
+        /// <summary>
+        /// Returns whether the given source type is compatible with the given destination type
+        /// based on the definition of "compatible-with" in the ECMA-335 standard (I.8.7.1).
+        /// </summary>
+        bool IsCompatibleWith(TypeDesc src, TypeDesc dst)
         {
-            if (src.Kind == dst.Kind && src.Type == dst.Type)
+            if (src == dst)
                 return true;
 
-            switch (src.Kind)
-            {
-            case StackValueKind.ObjRef:
-                if (dst.Kind != StackValueKind.ObjRef)
-                    return false;
-
-                // null is always assignable
-                if (src.Type == null)
-                    return true;
+            if (src == null || dst == null)
+                return false;
 
-                return CastingHelper.CanCastTo(src.Type, dst.Type);
+            // Everything is compatible to object through transitivity
+            if (dst.IsObject)
+                return true;
 
-            case StackValueKind.ValueType:
+            if (!src.IsPrimitive && src.CanCastTo(dst))
```

Comment:
What is the purpose of the `IsPrimitive` condition? It does not look right.

Reply:
I used the `CanCastTo` method to check for the transitivity and variance rules, as well as the remaining rules defined for reference types and arrays. However, CanCastTo would also return true for primitives that are assignable but not compatible.

Reply:
I think that this will cause the following to not to verify (compile with `/o+`):

```
static IComparable s;

static void foo(int x)
{
    s = x;
}
```

Reply:
That is true, but this is not actually caused by the IsPrimitive check, but by the current implementation of `box` creating a StackValue of StackValueKind `ObjRef` with the TypeDesc type of the boxed value. Therefore my new implementation of IsAssignable compares the boxed type as an Int32, as I did not account for this scenario yet.
Maybe it would be better to create additional IsAssignable overloads which accept StackValues and then call the TypeDesc variant with the appropriate TypeDesc (e.g. calling IsAssignable with a a type of System.Object, instead of the type of the boxed value). This would also most probably fix #4359.

Reply:
You are right. Investigating how boxing currently is implemented, lead me to understand why the IsAssignable method previously was implemented the way it was. I will have to rethink my assignability checks.